### DISCUSSION TITLE
Typegen and always events

### DISCUSSION THREAD

**Discussion by SimeonC** [statelyai/xstate#3195] [Q&A] [answered]
Wondering if it's OK to expect that "always" events should keep the same event types for typegen purposes. As I understand it "always" doesn't trigger another "event" it just changes the transition target - though I do get that this is probably quite complex to figure out.

It seems to work in practice even though the types say it shouldn't (not sure if this is intentional or I'm abusing APIs again).

See this [TS Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwFBzgYygUwIY0wWWzQAtgA7TJAXzgDMoIQ4AiADwGcZdNmBuJJJlaRYccnii0imOABUAnmEwBzTGQCMiFHADkAAT0cueAPQxFKtToBccGFACumfqkwA3NTHYBhbA-bkygCCaDDAEGTstghULmJkEmTYADYAoh4JUVqoqDo2iHYWtnlwsdq5RtwAdOTAMPkIhUrFlXg1ZHU6pXG5ACYRmO1uEADWgwBqKcC93A3lOebNuv0UQ6MTUzN4Oj05W9i2DmQjZBAA7mS7qAD612FgxQDKmDIwxDIAGo-GMrKPcP00Ow7BA4MkcFAyHBiOcQXBOAwyMpkvImq9SOwqjt5mUcrpMFAGFAqmBkrhaNAQFVJslprNovNUItMMUCUSSWSYBSoFSaXTtlcAbgDnAjidzpccXFcfFhmNHlA0AA5bAgAjYB7ZHJ8rYs5YDNZjambWbSuIgYDsAJIgCS4HBaoSuHCkQZeKIYQiWQoHigu3YBLcwDQmG97gJu2UDmwUF6WR0AAlsMCdc6Iti8b1MGT5GHfWbtOHMr5-IFnlAgyGssg8am8MUM91CxkvCXrcoAOLR2PV+aJ5NwOsu-J5AuuFs+PztgAi2ewueiMpAuBIoe+3Ci8wAPrpavVt7pZKGJDoDzoj5wCVVvO80CNT3id+fj1e65gHzknxeJFVpwMP6gO6NN+BIAPzFDemB3l0T5vjBuh-hQXQylwyh5hG1ACEgaBevAeCcIQJDkDIAC82gmAAVBRcCGJw3AALQ5hADjwEqAAsNrKGAygAFLKAArCA04API2t4-EAIpBBA3gAEwAELTkErDKN4Zw8QAmtO7BnPJADSQTCcoSq6Wg8nAEEABaEkAGJoEEABW4yPKwACqQTyWcCYwJZAAyaAAAoAJwAOpSdOAAMEnydgQSpLpslBEEZwjGAQQAErACM8kgMJjxuCiPFBOMwARUE2CpHFQXpWg8iPLIlXCUEHbyHpSr8bxrVBL5060MkjxKkEHzqJgABsEUhQFNrjDaxVNecQTeGgjzNTZ7D4PF5XyBA8myCAbF6R8ZBsSM002bI3jpfg+DAAAzPgSodhFvQAF42awMC9MlEkObISpKgm6UheMtXjEqNmYB8vQBdOnneBpxCzbJ3jThJyjTuMZXyTAaDeAmaMAI7KMoADU+DTjxAAcQTTmgwnYPgIAk2cmAdqTelsZgAXKDZxDyfJqQOakIy9LIwnjA9056WgABGGn0TANmpOw6U2fgHzsI8HZvS9DgSeoCbeBJmAvWAaVuOMMCU8J6neI8NrsKQ+BnDZEkSewqTCXpqSyQFEXpcEL1axAEmpNOGmpFAM08b0GlgK58kywmGnyB2ZyJWxlloGcjy1fxel7ZgZzYBzNrpRJ9V6bdmDyMJ52JSA-G9CzYdsdOygSU13gy2xJO9FAIzeMAUDpb5PGyNgwDO-I3iuc8rBuB2rCp6wgs8a5lOyAFgvJAFPFuNgrCyRJo0gLI-GU-YJO0KNxDeDx6XYL53gOTZwAAOyWSArnyPI6XpQmXynccq3TQG5YAJ1sBnFcuQEAvRHh6RCgTUaQUyCjVuqNXoMswA2jABpZQlN1A8WnBqGWb95KuRGK5Ly2BXIeHkLJVGjxhLpVcjCCqxBUhsUEtZW604EEBVclyEAHYGBBGCMAeKDlmIOUkcJEgL00AgEstOEABNUjKDQCTVyL15BOQtMJWQlpXJkFXsQXyt1jK0DOFAM44woCuUPhAYgDh8DKFkKNEYQU9Iy2EqNTA4x2Ay1kuMLilM1E2XcSMZICYAqyRtBFSmrBsAhVkg5NwsjIYExlh8EmZAbLd3GCMBy8glR6WyqwHiMslRoCknACiJhtAYBwHgQipAKAAAp5g1jxHYdgCglDVhoAOAZlgNAABpGTwlXMuN0vTxyeCGXAAcjRmRPAcDLC09ReBwAPlAYA2AEjFA9MADw8EdDkBOWc0oUyqCTPmXuA5yRih7h0Pc3pdF8JzIeR0GA3z5lwAiP8gFcBHgbK2cCkFdgYyqD+YeF8UAAJQqoFMvEKKQV3NRXIBFkK8SPJSBBW8953kgs+aGXFvTIJ3gpfMlIxcFxwAANpYpBT0qFAKcJkF6MUJMKYTSejIG8llUKuBQFhcUOCwr5mYvZfMtlsq8SivFQhf8UrenooVXAAAumqmVsq3w0t6eQOUep5UKvYIqCV-L3wks1dMK1tJdRCs1TkCIiE9TMpdXiM1XqlWYDhTod1SKXUapdVq21CqIiRyJLYT1XrUA+pdX6gNQa1XSrTTkHVIbdURoBe6xcwrQ3qtubm+1ugOnLiIhQAAlEivVCapkBgrMGclWp5kGrgB0zleAPrjLgEWGA1a4AkQAHxtoBcAWgnaB1VGZMOkiJFdBgs2Z0IdnL2AQHBFUZIEBlAdJnXsg5CRq2CjxFgGADhIRwACgwC0AYqhYA3ckDwHST1YqLagItGqT1AA) for reference.

**Comment by Andarist** [comment, marked as answer]
This is actually a bug in XState (sort of). At the moment this is mentioned in the [known limitations of typegen](https://xstate.js.org/docs/guides/typescript.html#typegen). The problem is that while this invoke creator might receive `SUBMIT` event, some actions and guards might actually receive `''` event. 

I've created a ticket internally last week to fix this. However, it's not a straightforward change and we need to decide if we want to actually do this in v4, cc @davidkpiano

**Reply by SimeonC** [reply]
Thanks, I keep forgetting to check the Known Limitations sections in the docs as well as the issues/discussions.
